agent/lib/kontena/helpers/observable.rb

```diff
@@ -0,0 +1,163 @@
+    # Updates to value will send to async method on given actor
+    # Returns current value
+    #
+    # @param actor [celluloid::Actor]
+    # @param method [Symbol]
+    # @return [Object, nil] possible existing value
+    def add_observer(observer, observe)
+      debug "observer: #{observer} <- #{@value.inspect[0..64] + '...'}"
+
+      observers[observer] = observe
+
+      return @value
+    end
+
+    def notify_observers
+      observers.each do |observer, observe|
+        begin
+          debug "notify: #{observer} <- #{@value}"
+
+          # XXX: is the Observable's Celluloid.current_actor guranteed to match the Actor[:node_info_worker] Celluloid::Proxy::Cell by identity?
+          observer.mailbox << Message.new(Celluloid.current_actor, observe, @value)
+        rescue Celluloid::DeadActorError => error
+          observers.delete(actor)
+        end
+      end
+    end
+  end
+
+  # An Actor that observes the value of other Obervable Actors
+  module Observer
+    include Kontena::Logging
+
+    # A task observing some Observables, and tracking their values
+    class Observe
+      # @param observables [Array<Observable>] Observable actors
+      # @param block [Block] Observing block
+      def initialize(observables, block)
+        @observables = observables
+        @block = block
+        @values = { }
+      end
+
+      # Set value for observable
+      #
+      # @return value
+      def set(observable, value)
+        @values[observable] = value
+      end
+
+      # Each observable has a value?
+      #
+      # @return [Boolean]
+      def ready?
+        !@values.any? { |observable, value| value.nil? }
+      end
+
+      # Map each observable to its value
+      #
+      # @return [Array] values or nil
+      def values
+        @observables.map{|observable| @values[observable] }
+      end
+
+      # Yield to block with values
+      def call
+        @block.call(*values)
+      end
+    end
+
+    # Yield values from Observables.
+    # Yields to block once all observed values are valid, and when they are updated.
+    # Crashes this Actor if any of the observed Actors crashes.
+    #
+    # Yields to block from new async task. Setup happens sync, and will raise on broken observables
+    #
+    # @param observables [Array{Observable}]
+    # @return [Observe]
+    # @yield [] all observables are valid
+    def observe(*observables, &block)
+      # unique handle to identify this observe loop
+      observe = Observe.new(observables, block)
+
+      # sync setup of each observable
+      observables.each do |observable|
+        value = observe.set(observable, observable.add_observer(Celluloid.current_actor, observe))
+
+        debug "observe #{observable} -> #{value}"
+
+        # crash if observed Actor crashes, otherwise we get stuck without updates
+        self.link observable
+      end
+
+      # async update message loop
+      Thread.current[:celluloid_actor].task :observe do
```

Comment:
This is based on how Celluloid implements e.g. timers, but maybe it's too much into the internals... Use `self.async._obseve_task(observe)` instead?